### PR TITLE
API overhaul and cleanup

### PR DESCRIPTION
**Pull request by zshipko**
Continues the work done on #11 with some additional fixes and cleanup

- Major API overhaul
  * Converted `Commit`, `Tree`, `Sync` and `Pack` to capabilities
- Update `Tree` API to avoid serializing trees as much as possible 
  * When using trees over RPC, the tree is now always stored on the server and any modification to it happens on the server side rather than passing entire trees back and forth between the server and client.
- Upgrade to Irmin 2.3
- Additional logging
- Added CHANGES.md and fixed CI

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,48 +1,22 @@
...
-RUN autoreconf -i
-RUN ./configure
-RUN make -j6 check
-RUN make install
-RUN which capnp
-
 # base
-FROM ocaml/opam2:alpine as base
-RUN sudo apk add --update m4 gmp gmp-dev perl
+FROM ocaml/opam2:debian-stable as base
```
Comment (avsm) [inline Dockerfile:2]:
```suggestion
FROM ocaml/opam as base
```

This should use `ocaml/opam` now, as ocaml/opam2 is being deprecated.

File: .ocamlformat [hunk 2]
```diff
@@ -1,3 +1,3 @@
-version = 0.14.1
+version = 0.15.0
```
Comment (CraigFe) [inline .ocamlformat:1]: Might as well bump to 0.16.0, while you're here.

File: examples/client.ml [hunk 3]
```diff
@@ -1,19 +1,38 @@
...
+let ( let* ) = Lwt.bind
 
 let main =
-  Rpc.Client.connect (Uri.of_string uri) >>= fun client ->
-  Rpc.Client.set client ~author ~message [ "abc" ] "123" >>= fun _ ->
-  Rpc.Client.get client [ "abc" ] >|= fun res ->
+  let* client = Uri.of_string uri |> Client.connect in
+  let* repo = Client.repo client in
+  let* master = Client.Store.master repo in
+  Client.ping client >>= fun x ->
```
Comment (CraigFe) [inline examples/client.ml:22]:
```suggestion
  let* x = Client.ping client in
```

File: examples/client.ml [hunk 4]
```diff
@@ -1,19 +1,38 @@
...
-  print_endline res
+  print_endline res;
+  let* tree = Client.Store.get_tree master [ "abc" ] in
+  let* concrete = Client.Tree.concrete tree in
+  assert (match concrete with `Contents _ -> true | _ -> false);
+  (*let* l = Store.Tree.list tree [] in
+    assert (List.length l = 1);*)
+  (*let* pack = Client.Store.pack master in
+    let* check = Client.Store.Pack.integrity_check (Option.get pack) in
+    assert (Result.is_ok check);*)
```
Comment (CraigFe) [inline examples/client.ml:35]: Stray comment here.

File: examples/mirage/_tags [hunk 5]
```diff
@@ -0,0 +1 @@
+true: package(digestif.ocaml checkseum.ocaml)
```
Comment (CraigFe) [inline examples/mirage/_tags:1]: This looks like it belongs in `.gitignore`?

File: src/irmin-rpc-unix/irmin_rpc_unix.mli [hunk 6]
```diff
@@ -35,7 +35,7 @@ module Make
   end
 
   module Client : sig
-    include Irmin_rpc.Client.S with module Store = Store
+    include module type of Irmin_rpc.Client.Make (Store) (Remote) (Pack)
```
Comment (CraigFe) [inline src/irmin-rpc-unix/irmin_rpc_unix.mli:38]: Worth avoiding `module type of` if it's possible here: we're providing `Client.S` for this reason.

File: src/irmin-rpc/client_intf.ml [hunk 7]
```diff
@@ -1,107 +1,237 @@
...
 
-  val merge :
-    t ->
-    ?branch:Store.branch ->
-    author:string ->
-    message:string ->
-    Store.branch ->
-    (Store.Hash.t, [ `Msg of string ]) result Lwt.t
+  type pack
+  (** Pack handle, user for accessing `Irmin_pack` functionality *)
```
Comment (CraigFe) [inline src/irmin-rpc/client_intf.ml:42]: These types are already included in `Types`; did you mean to shadow those here?

File: test/test.ml [hunk 8]
```diff
@@ -1,152 +1,140 @@
...
+        Client.get_tree master [ "k" ]
+        >>= Client.Tree.exists
+        >|= Alcotest.(check bool) "Tree exists" true
+        (*>|= Alcotest.(check find_tree) "Binding [k → Some tree]" (Some tree)*)
+      in*)
+    (*let* () =
+        Client.find_tree master [ "k_absent" ]
+        >>= Option.map_lwt Server.Tree.to_concrete
+        >|= Alcotest.(check find_tree) "Binding [k_absent → Some tree]" None
+      in*)
```
Comment (CraigFe) [inline test/test.ml:101]: Stray failing test?

**Review by CraigFe — APPROVED** [review]
LGTM.

I'm a little unhappy to see pack-specific functionality sneaking into the interface here – particularly if it's just for the purpose of integrity checking – since it further splits the codebase into Pack and non-Pack features. It doesn't look too hard to change later though, so let's go with it for now.

**Comment by samoht** [comment]
> When using trees over RPC, the tree is now always stored on the server and any modification to it happens on the server side rather than passing entire trees back and forth between the server and client

Curious on how you deal with lifetime of these values on the server-side: is there some potential to blowing the daemon memory away if these are never GCed?

**Comment by zshipko** [comment]
> Curious on how you deal with lifetime of these tree values on the server-side: is there some potential to blowing the daemon memory away if these are never GCed?

For now I'm using `Capnp_rpc_lwt.Capability.when_released` to remove it from a `Hashtbl` when the capability gets garbage collected - it is still possible for a client to create many trees and overload the memory, but they are at least being garbage collected when the client disconnects. The issue with using a cache that would evict trees automatically is that we don't want to remove a tree that a client is in the process of manipulating just because another client created a bunch of trees

The solution I'm thinking of implementing is a per-client limit on the number of trees in addition to an RPC method that enables a client to explicitly mark a tree for garbage collection. I know the GC method would be pretty straight forward but I haven't thought about how I would go about associating trees with a specific client yet.

**Comment by zshipko** [comment]
I'm going to merge this now, since it is growing pretty big. There are some more things I would like to clean up but will make additional PRs for subsequent work.

**Pull request merged**